docs/ALGORITHM.md

```diff
@@ -0,0 +1,91 @@
+# Finality Quorum Calculation Algorithm
+
+## Table of Contents
+
+1. [Overview](#1-overview)  
+2. [Inputs and Data Sources](#2-inputs-and-data-sources)  
+3. [Step‑by‑Step Algorithm](#3-step-by-step-algorithm)
+
+---
+
+## 1. Overview
+
+The Finality Gadget determines if a rollup block is finalized by 
+checking if a quorum of voting power has signed the block. This 
+involves fetching votes from the contract, calculating each 
+provider’s voting power, and comparing the total voted power to 
+the required threshold (> 2/3).
+
+---
+
+## 2. Inputs and Data Sources
+
+1. Rollup blocks  
+   - Fetch height, hash, and timestamp from the rollup chain  
+2. Finality Providers  
+   - Fetch FP public keys from the Babylon chain  
+3. Voting Power  
+   - Calculate FP voting power via Bitcoin delegation data at the 
+     specific BTC block height  
+4. Quorum Threshold  
+   - Hardcoded as 2/3 of total voting power  
+
+---
+
+## 3. Step‑by‑Step Algorithm
+
+1. **Fetch activation parameters**  
+   Read `bsn_activation_height` and `finality_signature_interval` from the Rollup BSN contract.
+
+2. **Determine start block**  
+   ```pseudo
+   if currentBlockHeight < bsn_activation_height:
+       currentBlockHeight = bsn_activation_height
+   ```
```

Comment:
Actually if `currentBlockHeight < bsn_activation_height`, the algorithm can simply return because this block is not "finalise-able". Wdyt?